File: gray-scott/gs.py
import matplotlib.pyplot as plt
import numpy as np
from numba import jit


@jit
def laplacian(m, n, s):
    ts = 0.0
    ts += s[m + 1][n]
    ts += s[m - 1][n]
    ts += s[m][n + 1]
    ts += s[m][n - 1]
    ts -= 4.0 * s[m][n]
    return ts


@jit
def calc(u, v, u2, v2):
    (L, _) = u.shape
    dt = 0.2
    F = 0.04
    k = 0.06075
    Du = 0.1
    Dv = 0.05
    lu = np.zeros((L, L))
    lv = np.zeros((L, L))
    for ix in range(1, L - 1):
        for iy in range(1, L - 1):
            lu[ix, iy] = Du * laplacian(ix, iy, u)
            lv[ix, iy] = Dv * laplacian(ix, iy, v)
    cu = -v * v * u + F * (1.0 - u)
    cv = v * v * u - (F + k) * v
    u2[:] = u + (lu + cu) * dt
    v2[:] = v + (lv + cv) * dt


@jit
def simulation(L, loop):
    u = np.zeros((L, L))
    u2 = np.zeros((L, L))
    v = np.zeros((L, L))
    v2 = np.zeros((L, L))
    h = L // 2
    u[h - 6:h + 6, h - 6:h + 6] = 0.9
    v[h - 3:h + 3, h - 3:h + 3] = 0.7
    r = []
    for i in range(loop):
        calc(u, v, u2, v2)
        u, u2, v, v2 = u2, u, v2, v
        if i % 100 == 0:
            r.append(v.copy())
    return r


def save_imgs(imgs):
    n = len(imgs)
    for i in range(4):
        filename = f"conf{i:04}.png"
        plt.imsave(filename, imgs[n // 4 * i])


def save_as_vtk(filename, img):
    f = open(filename, "w")
    print(filename)
    L = len(img)
    f.write("# vtk DataFile Version 1.0\n")
    f.write(f"{filename}\n")
    f.write("ASCII\n")
    f.write("DATASET UNSTRUCTURED_GRID\n")
    f.write(f"POINTS {img.size} float\n")
    for i in range(L):
        for j in range(L):
            f.write(f"{i} {j} 0\n")
    f.write(f"CELLS {img.size} {img.size*2}\n")
    for i in range(img.size):
        f.write(f"1 {i}\n")
    f.write(f"CELL_TYPES {img.size}\n")
    for _ in range(img.size):
        f.write(f"1\n")
    f.write(f"POINT_DATA {img.size}\n")
    f.write("SCALARS scalars float\n")
    f.write("LOOKUP_TABLE default\n")
    img = img.reshape(L * L)
    for i in range(img.size):
        f.write(f"{img[i]:.03f}\n")


imgs = simulation(80, 1000)

for i in range(len(imgs)):
    save_as_vtk(f"conf{i:03}.vtk", imgs[i])

# imgs = simulation(80, 15000)
# save_imgs(imgs)

File: gray-scott/gs.cpp
#include <fstream>
#include <iomanip>
#include <iostream>

const int L = 80;
const double F = 0.04;
const double k = 0.06075;
const double dt = 0.2;
const double Du = 0.05;
const double Dv = 0.1;

void init(double u[L][L], double v[L][L]) {
  int d = 3;
  for (int i = L / 2 - d; i < L / 2 + d; i++) {
    for (int j = L / 2 - d; j < L / 2 + d; j++) {
      u[i][j] = 0.7;
    }
  }
  d = 6;
  for (int i = L / 2 - d; i < L / 2 + d; i++) {
    for (int j = L / 2 - d; j < L / 2 + d; j++) {
      v[i][j] = 0.9;
    }
  }
}

double calcU(double tu, double tv) {
  return tu * tu * tv - (F + k) * tu;
}

double calcV(double tu, double tv) {
  return -tu * tu * tv + F * (1.0 - tv);
}

double laplacian(int ix, int iy, double s[L][L]) {
  double ts = 0.0;
  ts += s[ix - 1][iy];
  ts += s[ix + 1][iy];
  ts += s[ix][iy - 1];
  ts += s[ix][iy + 1];
  ts -= 4.0 * s[ix][iy];
  return ts;
}

void calc(double u2[L][L], double v2[L][L], double u[L][L], double v[L][L]) {
  for (int ix = 1; ix < L - 1; ix++) {
    for (int iy = 1; iy < L - 1; iy++) {
      double du = 0;
      double dv = 0;
      du = Du * laplacian(ix, iy, u);
      dv = Dv * laplacian(ix, iy, v);
      du += calcU(u[ix][iy], v[ix][iy]);
      dv += calcV(u[ix][iy], v[ix][iy]);
      u2[ix][iy] = u[ix][iy] + du * dt;
      v2[ix][iy] = v[ix][iy] + dv * dt;
    }
  }
}

void save_as_vtk(double u[L][L]) {
  static int index = 0;
  char filename[256];
  sprintf(filename, "conf%03d.vtk", index);
  std::cerr << filename << std::endl;
  index++;
  std::ofstream ofs(filename);
  ofs << "# vtk DataFile Version 1.0" << std::endl;
  ofs << filename << std::endl;
  ofs << "ASCII" << std::endl;
  ofs << "DATASET UNSTRUCTURED_GRID" << std::endl;
  ofs << "POINTS " << L * L << " float" << std::endl;

  for (int i = 0; i < L; i++) {
    for (int j = 0; j < L; j++) {
      ofs << i << " " << j << " 0" << std::endl;
    }
  }

  ofs << "CELLS " << L * L << " " << L * L * 2 << std::endl;
  for (int i = 0; i < L * L; i++) {
    ofs << "1 " << i << std::endl;
  }

  ofs << "CELL_TYPES " << L * L << std::endl;
  for (int i = 0; i < L * L; i++) {
    ofs << "1" << std::endl;
  }

  ofs << "POINT_DATA " << L * L << std::endl;
  ofs << "SCALARS scalars float" << std::endl;
  ofs << "LOOKUP_TABLE default" << std::endl;
  ofs << std::fixed;
  for (int i = 0; i < L; i++) {
    for (int j = 0; j < L; j++) {
      ofs << u[i][j] << std::endl;
    }
  }
  ofs.close();
}

int main(void) {
  double u[L][L] = {};
  double v[L][L] = {};
  double u2[L][L] = {};
  double v2[L][L] = {};
  init(u, v);
  for (int i = 0; i < 12000; i++) {
    if (i % 2 == 0) {
      calc(u2, v2, u, v);
    } else {
      calc(u, v, u2, v2);
    }
    if (i % 100 == 0) {
      save_as_vtk(u);
    }
  }
}
